libraries/AP_AccelCal/AccelCalibrator.cpp

```diff
@@ -0,0 +1,464 @@
+{
+    clear();
+}
+/*
+    Select options, initialise variables and initiate accel calibration
+    Options:
+    Fit Type:       Will assume that if accelerometer static samples around all possible orientatio
+                    are spread in space will cover a surface of AXIS_ALIGNED_ELLIPSOID or any general 
+                    ELLIPSOID as chosen
+
+    Num Samples:    Number of samples user should will be gathering, please note that with type of surface
+                    chosen the minimum number of samples required will vary, for instance in the case of AXIS
+                    ALIGNED ELLIPSOID atleast 6 distinct samples are required while for generic ELLIPSOIDAL fit
+                    which will include calculation of offdiagonal parameters too requires atleast 8 parameters
+                    to form a distinct ELLIPSOID
+
+    Sample Time:    Time over which the samples will be gathered and averaged to add to a single sample for least
+                    square regression
+
+    offset,diag,offdiag: initial parameter values for LSQ calculation
+*/
+void AccelCalibrator::start(enum accel_cal_fit_type_t fit_type, uint8_t num_samples, float sample_time) {
+    start(fit_type, num_samples, sample_time, Vector3f(0,0,0), Vector3f(1,1,1), Vector3f(0,0,0));
+}
+
+void AccelCalibrator::start(enum accel_cal_fit_type_t fit_type, uint8_t num_samples, float sample_time, Vector3f offset, Vector3f diag, Vector3f offdiag) {
+    if (_status == ACCEL_CAL_FAILED || _status == ACCEL_CAL_SUCCESS) {
+        clear();
+    }
+    if (_status != ACCEL_CAL_NOT_STARTED) {
+        return;
+    }
+
+    _conf_num_samples = num_samples;
+    _conf_sample_time = sample_time;
+    _conf_fit_type = fit_type;
+
+    _param_struct.offset = offset;
+    _param_struct.diag = diag;
+    _param_struct.offdiag = offdiag;
+
+    switch (_conf_fit_type) {
+        case ACCEL_CAL_AXIS_ALIGNED_ELLIPSOID:
+            if (_conf_num_samples < 6) {
+                set_status(ACCEL_CAL_FAILED);
+                return;
+            }
+            break;
+        case ACCEL_CAL_ELLIPSOID:
+            if (_conf_num_samples < 8) {
+                set_status(ACCEL_CAL_FAILED);
+                return;
+            }
+            break;
+    }
+
+    set_status(ACCEL_CAL_WAITING_FOR_ORIENTATION);
+}
+
+
+// set Accel calibrator status to make itself ready for future accel cals
+void AccelCalibrator::clear() {
+    set_status(ACCEL_CAL_NOT_STARTED);
+}
+
+// returns true if accel calibrator is running
+bool AccelCalibrator::running() {
+    return _status == ACCEL_CAL_WAITING_FOR_ORIENTATION || _status == ACCEL_CAL_COLLECTING_SAMPLE;
+}
+
+// set Accel calibrator to start collecting samples in the next cycle
+void AccelCalibrator::collect_sample() {
+    set_status(ACCEL_CAL_COLLECTING_SAMPLE);
+}
+
+// collect and avg sample to be passed onto LSQ estimator after all requisite orientations are done
+void AccelCalibrator::new_sample(Vector3f delta_velocity, float dt) {
```

Comment:
please pass vectors as const references where possible, or it just wastes time and stack space